### PR TITLE
Automatic conversion from single to multi geometry types

### LINKED ISSUE #191: Rust writer: Mixed Polygon/Multipolygon input results in invalid FGB output
Converting `countries.geojson` gives an error in GDAL:

```
ogrinfo -al test_multipoly.fgb
...
OGRFeature(countries):150
  name (String) = Paraguay
  POLYGON ((-62.685057 -22.249029,-62.291179 -21.051635,-62.265961 -20.513735,-61.786326 -19.633737,-60.043565 -19.342747,-59.115042 -19.356906,-58.183471 -19.868399,-58.166392 -20.176701,-57.870674 -20.732688,-57.937156 -22.090176,-56.88151 -22.282154,-56.473317 -22.0863,-55.797958 -22.35693,-55.610683 -22.655619,-55.517639 -23.571998,-55.400747 -23.956935,-55.027902 -24.001274,-54.652834 -23.839578,-54.29296 -24.021014,-54.293476 -24.5708,-54.428946 -25.162185,-54.625291 -25.739255,-54.788795 -26.621786,-55.695846 -27.387837,-56.486702 -27.548499,-57.60976 -27.395899,-58.618174 -27.123719,-57.63366 -25.603657,-57.777217 -25.16234,-58.807128 -24.771459,-60.028966 -24.032796,-60.846565 -23.880713,-62.685057 -22.249029))

ERROR 1: Unexpected nullptr: XY data
ERROR 1: Failed to read geometry
ERROR 1: Fatal error parsing feature
```

Failing feature is:
```
{"type":"Feature","id":"QAT","properties":{"name":"Qatar"},"geometry":{"type":"Polygon","coordinates":[[[50.810108,24.754743],[50.743911,25.482424],[51.013352,26.006992],[51.286462,26.114582],[51.589079,25.801113],[51.6067,25.21567],[51.389608,24.627386],[51.112415,24.556331],[50.810108,24.754743]]]}}
```

Expected output:
```
OGRFeature(countries):134
  name (String) = Qatar
  MULTIPOLYGON (((50.810108 24.754743,50.743911 25.482424,51.013352 26.006992,51.286462 26.114582,51.589079 25.801113,51.6067 25.21567,51.389608 24.627386,51.112415 24.556331,50.810108 24.754743)))
```

### PR DESCRIPTION
**Pull request by pka**
Fixes #191

### CODE REVIEW

File: src/rust/src/file_writer.rs [hunk 1]
```diff
@@ -84,7 +84,7 @@ impl<'a> FgbWriter<'a> {
 
         cfgfn(&mut fbb, &mut header_args);
 
-        let mut feat_writer = FeatureWriter::new();
+        let mut feat_writer = FeatureWriter::new(header_args.geometry_type, true);
```
Comment (michaelkirk) [inline src/rust/src/file_writer.rs:87]: I'm surprised by defaulting `force_multi` to true here. For example, that's the opposite of what ogr2ogr does, right?

Should this be user configurable, maybe via the cfgfn?

**Review by michaelkirk — COMMENTED** [review]
Seems like a useful feature!

File: src/rust/src/feature_writer.rs [hunk 2]
```diff
@@ -23,6 +23,10 @@ pub struct FeatureWriter<'a> {
     tm: Vec<u64>,
     // Type of geometry (only relevant for elements in heterogeneous collection types)
     type_: GeometryType,
+    // Declared geometry type of dataset (Unknown, GeometryCollection or well-defined type)
+    pub(crate) dataset_type: GeometryType,
+    // Convert single to multi geometries
+    force_multi: bool,
```
[the remark this answers is not in the document]
Comment (pka) [reply src/rust/src/feature_writer.rs:29]: Yes, it's the same, I think.

**Comment by pka** [comment]
To adress #209, I will write feature geometry types only when necessary and add options for "promote-to-multi" and also "detect-type".

**Comment by pka** [comment]
Opinions especially welcome for the new Writer API in 1be9bc9 @michaelkirk @bjornharrtell

**Review by bjornharrtell — APPROVED** [review]
Looks sensible to me.

File: src/rust/src/feature_writer.rs [hunk 3]
```diff
@@ -60,19 +71,90 @@ impl<'a> FeatureWriter<'a> {
...
+                    let type_ = match geometry_type {
+                        GeometryType::LineString if self.promote_to_multi => {
+                            self.geom_state = GeomState::ForceMulti;
+                            GeometryType::MultiLineString
+                        }
+                        GeometryType::Polygon if self.promote_to_multi => {
+                            self.geom_state = GeomState::ForceMulti;
+                            GeometryType::MultiPolygon
+                        }
+                        _ => geometry_type,
```
Comment (michaelkirk) [inline src/rust/src/feature_writer.rs:97]: I see the promotions for MultiLineString and MultiPolygon - do we need a similar promotion for MultiPoint?

**Review by michaelkirk — COMMENTED** [review]
Sorry for the slow review!

File: src/rust/src/feature_writer.rs (hunk 3 above)
Comment (pka) [reply src/rust/src/feature_writer.rs:97]: Probably yes. It's such a rare case that I forgot about.

**Pull request merged**